modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/RebalanceRaftGroupEventsListener.java

```diff
@@ -252,92 +294,189 @@ private void scheduleChangePeers(List<PeerId> peers, long term) {
 
-                    doOnNewPeersConfigurationApplied(peers);
-                }
+            List<ClusterNode> retrievedSwitchReduce = readClusterNodes(switchReduceEntry);
+            List<ClusterNode> retrievedSwitchAppend = readClusterNodes(switchAppendEntry);
+            List<ClusterNode> retrievedStable = readClusterNodes(stableEntry);
 
-                LOG.info("Rebalance finished. Going to schedule next rebalance [partition={}, table={}, appliedPeers={}, plannedPeers={}]",
-                        partNum, tblConfiguration.name().value(), appliedPeers, ByteUtils.fromBytes(plannedEntry.value()));
-            } else {
-                if (!metaStorageMgr.invoke(If.iif(
-                        notExists(plannedPartAssignmentsKey(partId)),
-                        ops(put(stablePartAssignmentsKey(partId), ByteUtils.toBytes(appliedPeers)),
-                                remove(pendingPartAssignmentsKey(partId))).yield(true),
-                        ops().yield(false))).get().getAsBoolean()) {
-                    LOG.info("Planned key changed while trying to update rebalance information. Going to retry"
-                                    + " [key={}, partition={}, table={}, appliedPeers={}]",
-                            plannedPartAssignmentsKey(partId), partNum, tblConfiguration.name(), appliedPeers);
+            // Were reduced
+            List<ClusterNode> reducedNodes = subtract(retrievedSwitchReduce, stable);
 
-                    doOnNewPeersConfigurationApplied(peers);
-                }
+            // Were added
+            List<ClusterNode> addedNodes = subtract(stable, retrievedStable);
 
-                LOG.info("Rebalance finished [partition={}, table={}, appliedPeers={}]",
-                        partNum, tblConfiguration.name().value(), appliedPeers);
-            }
+            // For further reduction
+            List<ClusterNode> calculatedSwitchReduce = subtract(retrievedSwitchReduce, reducedNodes);
 
-            rebalanceAttempts.set(0);
-        } catch (InterruptedException | ExecutionException e) {
-            // TODO: IGNITE-14693
-            LOG.warn("Unable to commit partition configuration to metastore [table = {}, partition = {}]",
-                    e, tblConfiguration.name(), partNum);
-        }
-    }
+            // For further addition
+            List<ClusterNode> calculatedSwitchAppend = union(retrievedSwitchAppend, reducedNodes);
+            calculatedSwitchAppend = subtract(calculatedSwitchAppend, addedNodes);
+            calculatedSwitchAppend = intersect(calculatedAssignments, calculatedSwitchAppend);
 
-    private static List<ClusterNode> resolveClusterNodes(
-            List<PeerId> peers, byte[] pendingAssignments, byte[] stableAssignments) {
-        Map<NetworkAddress, ClusterNode> resolveRegistry = new HashMap<>();
+            var calculatedPendingReduction = subtract(stable, retrievedSwitchReduce);
```

Comment:
What's up with all these "var"? We shouldn't use them here

Reply:
Right, forgot to clean them up